pkg/agent/agentserver/server_test.go

```diff
@@ -67,3 +81,147 @@ func TestAddRemoveBackends(t *testing.T) {
+	stub := gomock.NewController(t)
+	defer stub.Finish()
+
+	ns := "test_ns"
+	sa := "test_sa"
+
+	testCases := []struct {
+		desc               string
+		mdKey              string
+		tokens             []string
+		wantNamespace      string
+		wantServiceAccount string
+		authenticated      bool
+		authError          string
+		tokenReviewError   error
+		wantError          bool
+	}{
+		{
+			desc:      "no context",
+			wantError: true,
+		},
+		{
+			desc:      "non valid metadata key",
+			mdKey:     "someKey",
+			tokens:    []string{"token1"},
+			wantError: true,
+		},
+		{
+			desc:      "non valid token prefix",
+			mdKey:     header.AuthenticationTokenContextKey,
+			tokens:    []string{"token1"},
+			wantError: true,
+		},
+		{
+			desc:      "multiple valid tokens",
+			mdKey:     header.AuthenticationTokenContextKey,
+			tokens:    []string{header.AuthenticationTokenContextSchemePrefix + "token1", header.AuthenticationTokenContextSchemePrefix + "token2"},
+			wantError: true,
+		},
+		{
+			desc:               "not authenticated",
+			authenticated:      false,
+			mdKey:              header.AuthenticationTokenContextKey,
+			tokens:             []string{header.AuthenticationTokenContextSchemePrefix + "token1"},
+			wantNamespace:      ns,
+			wantServiceAccount: sa,
+			wantError:          true,
+		},
+		{
+			desc:               "tokenReview error",
+			authenticated:      false,
+			mdKey:              header.AuthenticationTokenContextKey,
+			tokens:             []string{header.AuthenticationTokenContextSchemePrefix + "token1"},
+			tokenReviewError:   fmt.Errorf("some error"),
+			wantNamespace:      ns,
+			wantServiceAccount: sa,
+			wantError:          true,
+		},
+		{
+			desc:               "non valid namespace",
+			authenticated:      true,
+			mdKey:              header.AuthenticationTokenContextKey,
+			tokens:             []string{header.AuthenticationTokenContextSchemePrefix + "token1"},
+			wantNamespace:      "_" + ns,
+			wantServiceAccount: sa,
+			wantError:          true,
+		},
+		{
+			desc:               "non valid service account",
+			authenticated:      true,
+			mdKey:              header.AuthenticationTokenContextKey,
+			tokens:             []string{header.AuthenticationTokenContextSchemePrefix + "token1"},
+			wantNamespace:      ns,
+			wantServiceAccount: "_" + sa,
+			wantError:          true,
+		},
+		{
+			desc:               "authorization succeed",
+			authenticated:      true,
+			mdKey:              header.AuthenticationTokenContextKey,
+			tokens:             []string{header.AuthenticationTokenContextSchemePrefix + "token1"},
+			wantNamespace:      ns,
+			wantServiceAccount: sa,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			kcs := k8sfake.NewSimpleClientset()
+
+			kcs.AuthenticationV1().(*fakeauthenticationv1.FakeAuthenticationV1).Fake.PrependReactor("*", "*", func(action k8stesting.Action) (handled bool, ret runtime.Object, err error) {
```

Comment:
For the first two parameters of PrependReactor, can you refine the scope? I guess they should be "create" and "tokenReviews".

Reply:
done